### ISSUE TITLE
investigate performance regressions in v0.2.0 + DyNet benchmarks

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#2518]
According to the scripts given in https://twitter.com/kchonyc/status/899832376955744257 there are two issues:
1. 0.2.0 has a 2x perf regression over 0.1.2 on small workloads
2. pytorch is ~10x slower than DyNet on this workload.

An issue to track this.

Here's the google doc with links to the code snippets, current numbers (with and without threading) and will update further as I investigate: https://docs.google.com/spreadsheets/d/1_k_mcxgR94OXH4qD6YYZNiS25QnKbQqAyB72KVD-Z9M/edit?usp=sharing

#### On PyTorch vs DyNet

So far, I've done stack sampling using `perf top` and it looks like MKL is spending significant time in barriers / threading overhead. DyNet on the other hand seems to be using Eigen's gemm for the workload.

#### On 0.2.0 vs 0.1.2
investigating, will edit comment once I know more.

Thanks to @kyunghyuncho @hal3 and @neubig for bringing this up. I will only close this issue once we're within reasonable ballpark of the DyNet numbers.

**Comment by neubig** [comment]
I'll be interested in hearing the reason for this when you figure out!

Also FYI: DyNet can also use MKL if you want to abstract away from differences in the backend: http://dynet.readthedocs.io/en/latest/install.html#mkl-support

**Comment by soumith** [comment]
Here's a full report on what's going on, what we are doing to fix this. We expect to finish the work in about 1.5 to 2 weeks and get pytorch to be within ballpark of DyNet numbers, work has been going on at full swing by @gchanan and @colesbury 

#### Reasons for slowdown

1. our autograd overhead is close to 10 microseconds per Function invocation.
2. with the introduction of higher-order gradients, this got worse in `0.2.0` than in `0.1.12`
3. minor CPU optimizations at places.

Turns out, Eigen is not a reason for the slowdown. We patched in Eigen into pytorch, and MKL into DyNet to verify this. Sorry about the initial false-positive.

#### Why didn't we catch it before-hand?

It might sound funny, but:
- we didn't realize folks are running scripts with really small workloads (no one told us)
- the smallest workload we run before every release for performance regressions is the small [word_language_model](https://github.com/pytorch/examples/tree/master/word_language_model), but it was not a small enough workload to catch these performance regressions.

#### How are we fixing it?

- We are doing work that's been long pending. All fundamental functions are being moved to our auto-generated [ATen](https://github.com/zdevito/aten) library, which is getting autograd support. Stuff like `torch.matmul`, `torch.sum`, etc.

The autograd overhead after doing this work will go from 10 microseconds to 1 microsecond or less.

We have evidence ([see gchanan's wiki below](https://github.com/gchanan/pytorch/wiki/Mini-Sequeence-Labeler-PyTorch-vs-Dynet)) that this will pessimistically get us to 2x of DyNet's speed on this workload, and optimistically, match DyNet's speed. I expect us to be within 1.5x of DyNet, while supporting all the existing features we have such as higher order derivatives. Once we reach 1.5x, we'll further investigate to close the gap to 1.0x. 

This work is independent of our `just-in-time` compiler that we are working on to accelerate small workloads by even more, by adding operator fusion and other optimizations.

#### Catching such regressions in the future

- We are setting up regression tests with smaller workloads (including this one).
- We will be working on deeply integrated profilers to easily identify bottlenecks in the future
- We will invite powerusers to throw their workloads at us and have dedicated engineering to investigate them routinely (more details on this later).

#### A more detailed, fuller report

@gchanan is maintaining a detailed wiki, with per-operator timing, etc. over here: https://github.com/gchanan/pytorch/wiki/Mini-Sequeence-Labeler-PyTorch-vs-Dynet
@neubig and @hal3 might be interested in this.

For @hal3 if you have some time to help us: [we couldn't reproduce the speed issues you had with the larger model](https://github.com/gchanan/pytorch/wiki/Mini-Sequeence-Labeler-PyTorch-vs-Dynet#larger-model). with your permission, I'll start an email chain with @gchanan so that he can reproduce the larger model issue as well, and investigate this further.

**Comment by apaszke** [comment]
There's an additional problem that seems to decrease our perf quite severely - dynamic CPU frequency management. The issue is that most work is done in short bursts in two threads (main Python thread and backward worker), and the `powersave` governor heavily downclocks their cores (~1.7GHz avg vs 2.6GHz)

With `powersave` mode, the script takes 16s on a development branch I'm on, but in `performance` it only takes 10s on my machine. To change CPU frequency governor run this command:
```bash
sudo cpupower frequency-set -g <powersave|performance>
```

I haven't checked how this affects DyNet run times. cc: @neubig

**Comment by jph00** [comment]
BTW if you're on Ubuntu, you can get cpupower through `sudo apt install linux-tools` .

**Comment by honnibal** [comment]
> Why didn't we catch it before-hand?

> It might sound funny, but:

> we didn't realize folks are running scripts with really small workloads (no one told us)
> the smallest workload we run before every release for performance regressions is the small > word_language_model, but it was not a small enough workload to catch these performance regressions.

I'm a little worried there's a misunderstanding here! Nobody is *really* running tasks with such small workloads --- the data from that Twitter thread was just a tiny sample, to ease discussion right?

I think there are tasks where the overhead becomes relevant, but not because the total job size is so small. For instance I expect a model that made a prediction on each character would be sort of slow in PyTorch?

**Comment by negrinho** [comment]
@honnibal, we are talking about 10x slower when compared to DyNet for the same model. This is prohibitively slow. There are cases in research where batching is difficult or you would rather bother with it later. In these cases, it is important to be able to get a working model that runs reasonably fast. This is is especially important in a dynamic frameworks such as PyTorch, as flexibility is one its main selling points. A factor of ten is the difference between getting results in an hour or getting results in ten hours. This definitely makes a difference in research projects where you want to iterate fast. 

There is always this idea that we are able to spend most of the time in fast C/C++ code, but I wonder if it really happens that frequently. I got uniformly slow Python code on more than one occasion. I'm definitely looking forward for the coming performance fix.

**Comment by honnibal** [comment]
@negrinho Maybe what I'm saying here is a bit subtle.

Certainly a difference between running in 10 hours and running in 1 hour matters a lot! If your experiments take 10h, you only really get to have one idea a day. That leaves a lot of time to get sad about the problem, and try to fix it :). Turning 10 minute tasks into 1 minute tasks is also very productive.

But turning 10s tasks into 1s tasks? I don't think that's intrinsically useful. It should only matter as evidence of a deeper problem, that would show up on the longer jobs.

I read @soumith 's comment as saying, "Oh, I didn't know people needed the library to scale down to 10s job sizes". My response is, I don't think anyone actually does! I wanted to check that effort wasn't being spent to optimise a micro-benchmark that doesn't matter.

**Comment by negrinho** [comment]
@honnibal, the overhead seems to be in the variable and backward computations, which are important parts of a training loop. If you do not batch, you end up getting a lot of these operations that perhaps spend most of their time on overhead. The way I read what @soumith said about workloads was not as much about the time they take, but rather about the amount of work that they you do. If you do not batch, the effective computation that you are doing per call may be tiny, which leads to overhead becoming an important issue. 

I believe these problem do come up in real implementations. The benchmark is a very minimal and straightforward example of something simple that is very inefficient. Right now, the only way of amortizing overhead seems to be doing everything using batches, which for some problems may involve a lot of tedious bookkeeping. For research, you often prefer quick and simple solutions at first.

**Comment by jph00** [comment]
I've seen this issue recently - the 10s becomes a lot more significant when you have a larger dataset, and need to run more epochs.

**Comment by hal3** [comment]
@soumith @gchanan just saw the note about the larger model size. is it worth touching base on this? happy to do it by email or here.

**Comment by soumith** [comment]
@hal3 let me open an email thread.

**Comment by negrinho** [comment]
@soumith any updates on this?

**Comment by gchanan** [comment]
@negrinho we've been working on porting a number of autograd ops to C++.  See https://github.com/pytorch/pytorch/pull/2805 for the initial work; I should have a PR up tomorrow that implements `torch.cat` in C++ as well.  Note that these implementations are currently (over)shadowed by the python definitions in variable.py; we want to do some more testing/development before we enable them by default.

With `addmm`, `add`, `transpose`, and `cat` implemented as C++ ops, we are seeing the mini sequence labeler PyTorch script with OMP_NUM_THREADS=1 running in ~5.5 seconds (down from ~8.7 seconds, see https://github.com/gchanan/pytorch/wiki/Mini-Sequence-Labeler-PyTorch-vs-Dynet for more details).  We are seeing speedups from moving each op to C++  approximately equal to the measured total backwards time for that op in python (these are listed on the wiki above); this is because we speed up the forwards as well so we are limited to just improving the backwards time.  So, adding up the time for the ops we still need to do should bring us to ~2.5-3.5 seconds total runtime.  This time is in line with the initial estimate of ~2.8 seconds we made before we started this process.

After we complete that, we have some ideas for what to optimize next, but need to do some more experiments/scoping before we know for sure what to target.  I'll keep this issue updated as we progress.

**Comment by khanhptnk** [comment]
Any update on this issue?

**Comment by gchanan** [comment]
We've continued to implement more operators as C++ autograd functions; when I checked last week, we were at ~3.5 seconds for the above benchmark.  I think indexing is the operation with the biggest impact that we still have to move.  @colesbury is working on that now.

**Comment by khanhptnk** [comment]
Thanks for the update. Great job! Looking forward to the release day!

**Comment by ChijunZhang** [comment]
What about now?

**Comment by gchanan** [comment]
I don't actively maintain the benchmark anymore, but the writeup is available at https://github.com/gchanan/pytorch/wiki/Mini-Sequence-Labeler-PyTorch-vs-Dynet.
